service/src/CpufreqSysfsDriver.cpp

```diff
@@ -0,0 +1,212 @@
+                    // The CPUs in this resource group span across multiple
+                    // indices in the current topology domain, so their
+                    // minimal common domain is at a more coarse level.
+                    is_correct_domain = false;
+                    break;
+                }
+            }
+            if (is_correct_domain && m_topo.is_nested_domain(m_domain, outer_domain)) {
+                // This cpufreq policy spans a domain that is more coarse
+                // than the most-coarse scope observed so far.
+                m_domain = outer_domain;
+                break;
+            }
+        }
+    }
+
+    int CpufreqSysfsDriver::domain_type(const std::string &) const
+    {
+        return m_domain;
+    }
+
+    std::string CpufreqSysfsDriver::attribute_path(const std::string &name,
+                                                   int domain_idx)
+    {
+        auto cpus_in_domain_idx = m_topo.domain_nested(GEOPM_DOMAIN_CPU, m_domain, domain_idx);
+
+        auto resource_it = M_CPUFREQ_RESOURCE_BY_CPU.end();
+        for (auto cpu_idx : cpus_in_domain_idx) {
+            resource_it = M_CPUFREQ_RESOURCE_BY_CPU.find(cpu_idx);
+            if (resource_it != M_CPUFREQ_RESOURCE_BY_CPU.end()) {
+                // Multiple CPUs may map to this domain_idx, but we only
+                // need to use the first mapping discovered.
+                break;
+            }
+        }
+        if (resource_it == M_CPUFREQ_RESOURCE_BY_CPU.end()) {
+            throw Exception("CpufreqSysfsDriver::attribute_path(): domain_idx " + std::to_string(domain_idx) + " does not have a cpufreq entry.",
+                            GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+
+        auto property_it = M_PROPERTIES.find(name);
+        if (property_it == M_PROPERTIES.end()) {
+            throw Exception("CpufreqSysfsDriver::attribute_path(): No such signal "
+                            + name,
+                            GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+
+        std::ostringstream oss;
+        oss << resource_it->second
+            << "/" << property_it->second.attribute;
+        return oss.str();
+    }
+
+    std::function<double(const std::string &)> CpufreqSysfsDriver::signal_parse(const std::string &signal_name) const
+    {
+        auto prop_it = M_PROPERTIES.find(signal_name);
+        if (prop_it == M_PROPERTIES.end()) {
+            throw Exception("CpufreqSysfsDriver::signal_parse(): Unknown signal name: " + signal_name,
+                            GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+        double scaling_factor = prop_it->second.scaling_factor;
+        return [scaling_factor](const std::string &content) {
+            if (content.find("<unsupported>") != content.npos) {
+                return static_cast<double>(NAN);
+            }
+            else {
+                return static_cast<double>(std::stoi(content) * scaling_factor);
```

Comment:
Add exception handling so that integer conversion results in a NAN in all cases, not just "<unsupported>".